ldms/src/ldmsd/ldmsd_config.c

```diff
@@ -161,7 +158,34 @@ ldmsd_plugin_t load_plugin(const char *plugin_name)
 	}
 	struct ldmsd_plugin *pi = pget();
 	if (pi)
-		pi->libpath = strdup(library_name);
+		*libpath_ = strdup(library_name);
+
+	/* Verify the following API contract:
+	 * - If flags & MULTI_INSTANCE is !0, then
+	 *   -- both the constructor and destructor must be defined
+	 *   -- term must _NOT_ be defined
+	 * - otherwise
+	 *   -- constructor and destructor must be NULL
+	 *   -- term may or may not be NULL
+	 */
+	if (pi->flags && LDMSD_PLUGIN_MULTI_INSTANCE) {
+		if (!pi->constructor || !pi->destructor || pi->term) {
+			ovis_log(config_log, OVIS_LERROR,
+				 "The library, '%s', has the multi-instance flag, "
+				 "but is missing either the constructor or "
+				 "destructor functions, or has term defined.\n",
+				 plugin_name);
+			goto err_0;
+		}
+	} else {
+		if (pi->constructor || pi->destructor) {
```

Comment:
I don't think we should do this. I don't think there is any reason to prevent single-instance plugins from using constructor() and destructor(). In fact we should _encourage_ their usage even for single-instance plugins so we can have the same API semantics for both single- mutli-instance plugins. After all, any single-instance plugin that we move away from term() is going to need to use destructor(). And if they use destructor they might want constructor again. 

Really, the only differentiator between single- and multi-instance plugins should be the flag.  ldmsd will stop a plugin without the LDMSD_PLUGIN_MULTI_INSTANCE from being loaded twice. Otherwise, everything else should be the same.

Reply:
I'm ok with this. But for the avoidance of doubt, if it *is* MULTI_INSTANCE, then constructor and destructor must be defined?

I do think we have to have some kind of API contract, e.g., if it is MULTI_INSTANCE, then constructor and destructor must be defined and term cannot be defined.

Reply:
Yes, I am fine with that rule for MULTI_INSTANCE. Thanks!